provisioning/service/src/HttpDeviceRegistrationClient.cs

```diff
@@ -0,0 +1,660 @@
+        {
+            return this.UpdateEnrollmentAsync(enrollment, CancellationToken.None);
+        }
+
+        public override Task<Enrollment> UpdateEnrollmentAsync(Enrollment enrollment, CancellationToken cancellationToken)
+        {
+            return this.UpdateEnrollmentAsync(enrollment, false, cancellationToken);
+        }
+
+        public override Task<Enrollment> UpdateEnrollmentAsync(Enrollment enrollment, bool forceUpdate)
+        {
+            return this.UpdateEnrollmentAsync(enrollment, forceUpdate, CancellationToken.None);
+        }
+
+        public override Task<Enrollment> UpdateEnrollmentAsync(Enrollment enrollment, bool forceUpdate, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+            ValidateRegistrationAndDeviceId(enrollment);
+
+            if (string.IsNullOrWhiteSpace(enrollment.ETag) && !forceUpdate)
+            {
+                throw new ArgumentException(ApiResources.ETagNotSetWhileUpdatingEnrollment);
+            }
+
+            var errorMappingOverrides = new Dictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>
+            {
+                {
+                    HttpStatusCode.Conflict,
+                    responseMessage => Task.FromResult((Exception) new EnrollmentAlreadyExistsException(enrollment.RegistrationId))
+                },
+                {
+                    HttpStatusCode.PreconditionFailed,
+                    async (responseMessage) => new PreconditionFailedException(await ExceptionHandlingHelper.GetExceptionMessageAsync(responseMessage))
+                }
+            };
+
+            PutOperationType operationType = forceUpdate ? PutOperationType.ForceUpdateEntity : PutOperationType.UpdateEntity;
+
+            return this.httpClientHelper.PutAsync(GetEnrollmentUri(enrollment.RegistrationId), enrollment, operationType, errorMappingOverrides, cancellationToken);
+        }
+
+        public override Task<BulkOperationResult> UpdateEnrollmentsAsync(IEnumerable<Enrollment> enrollments)
+        {
+            return this.UpdateEnrollmentsAsync(enrollments, false, CancellationToken.None);
+        }
+
+        public override Task<BulkOperationResult> UpdateEnrollmentsAsync(IEnumerable<Enrollment> enrollments, CancellationToken cancellationToken)
+        {
+            return this.UpdateEnrollmentsAsync(enrollments, false, cancellationToken);
+        }
+
+        public override Task<BulkOperationResult> UpdateEnrollmentsAsync(IEnumerable<Enrollment> enrollments, bool forceUpdate)
+        {
+            return this.UpdateEnrollmentsAsync(enrollments, forceUpdate, CancellationToken.None);
+        }
+
+        public override Task<BulkOperationResult> UpdateEnrollmentsAsync(IEnumerable<Enrollment> enrollments, bool forceUpdate, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+
+            foreach (Enrollment enrollment in enrollments)
```

Comment:
Duplicate code: we could extract a method and use this wherever we need to validate an `IEnumerable<Enrollment>` parameter.